### PR TITLE
Large json serialization/deserialization tests

### PR DESCRIPTION
**Pull request by buyaa-n**
Tought its better to use large object close to real world example. So used BlogSpot json class from @layomia's Real world tests PR dotnet#39094.
Added 2 tests: one for large size object producing json with size from 6kB to 15MB,  another for an object with deep nested structure up to max supported depth.

### CODE REVIEW

File: src/System.Text.Json/tests/Serialization/Stream.WriteTests.cs [hunk 1]
```diff
@@ -258,6 +259,488 @@ public static async Task LargeJsonFile(int bufferSize)
...
+                            The7Ef192B7F545 = new MentionedUser
+                            {
+                                UserId = $"7ef192b7f545",
+                                Username = "fagnerbrack",
+                                Name = "Fagner Brack",
+                                CreatedAt = 1456628553936, // TODO should be datetime
+                                ImageId = $"{i}*h8Ph7pgNeQHZTiLlah3h-A.jpeg",
+                                BackgroundImageId = string.Empty,
+                                Bio = "I believe ideas should be open and free. This is a non-profit initiative to write about fundamentals you won't find anywhere else. ~7 min post every few weeks.",
+                                TwitterScreenName = "FagnerBrack",
```
Comment (bartonjs) [inline src/System.Text.Json/tests/Serialization/Stream.WriteTests.cs]: I don't think we should be doing things that attribute to real-world entities.  All of the things about users should be made up.

File: src/System.Text.Json/tests/Serialization/Stream.WriteTests.cs [hunk 2]
```diff
@@ -258,6 +259,488 @@ public static async Task LargeJsonFile(int bufferSize)
...
+        [InlineData(4, false, false)]
+        [InlineData(8, true, true)]
+        [InlineData(8, true, false)]
+        [InlineData(8, false, true)]
+        [InlineData(8, false, false)]
+        [InlineData(16, true, true)]
+        [InlineData(16, true, false)]
+        [InlineData(16, false, true)]
+        [InlineData(16, false, false)]
+        public static async Task DeepJsonFileTest(int depthFactor, bool ignoreNull, bool writeIndented) {
```
Comment (bartonjs) [inline src/System.Text.Json/tests/Serialization/Stream.WriteTests.cs]: Style: Should be Allman braces, not K&R braces.

```suggestion
        public static async Task DeepJsonFileTest(int depthFactor, bool ignoreNull, bool writeIndented)
        {
```

File: src/System.Text.Json/tests/Serialization/BlogPost.cs [hunk 3]
```diff
@@ -0,0 +1,821 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Collections.Generic;
+
+namespace System.Text.Json.Serialization.Tests
```
Comment (bartonjs) [inline src/System.Text.Json/tests/Serialization/BlogPost.cs]: `Tag` and `Post` and `Metadata` and such are very generic terms, and might be unfortunate to mix into this namespace.

Consider making this file a sub-namespace, like `System.Text.Json.Serialization.Tests.Schemas.BlogPost`

**Comment by ericstj** [comment]
Fixes https://github.com/dotnet/corefx/issues/39500

File: src/System.Text.Json/tests/Serialization/BlogPost.cs [hunk 4]
```diff
@@ -0,0 +1,822 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Collections.Generic;
+using System.Net;
+
+namespace System.Text.Json.Serialization.Tests.Schemas.BlogPost
+{ 
+    public class Post
```
Comment (buyaa-n) [inline src/System.Text.Json/tests/Serialization/BlogPost.cs]: Talked with @layomia about this POCO, he used this generator https://app.quicktype.io/#l=cs&r=json2csharp which doc : https://github.com/quicktype/quicktype/blob/master/FAQ.md says: 
image 
so I assume there is no ownership/distribution/licensing issue with using this POCO.  
cc @richlander

File: src/System.Text.Json/tests/Serialization/Stream.WriteTests.cs [hunk 5]
```diff
@@ -258,6 +263,639 @@ public static async Task LargeJsonFile(int bufferSize)
...
+        [Theory]
+        [InlineData(128)]
+        [InlineData(1024)]
+        [InlineData(4096)]
+        [InlineData(8192)]
+        [InlineData(16384)]
+        [InlineData(65536)]
+        public static async void FlushThresholdTest(int bufferSize)
+        {
+            int thresholdSize = (int )(bufferSize * 0.9 - 2);
```
Comment (buyaa-n) [inline src/System.Text.Json/tests/Serialization/Stream.WriteTests.cs]: `bufferSize * 0.9` is threshold size from codebase - 2 for `["` characters, after these chars we add that amount of char `a` which produces buffer written equal to threshold size, after this long string adding a big object to produce scenario noted in: https://github.com/dotnet/corefx/issues/39500 :

> It would also be good to add tests where depth changes at those size threshold (for example, writing a start or end array/object at the 16k bytes)

File: src/System.Text.Json/tests/Serialization/BlogPost.cs (hunk 4 above)
Comment (ahsonkhan) [reply src/System.Text.Json/tests/Serialization/BlogPost.cs]: I learned recently from @terrajobst, there is a way to generate .NET POCO from JSON built-into Visual Studio. I don't know how accurate it is, but we could consider using that instead.

image


Doing so from the json file in https://github.com/dotnet/corefx/pull/39094 (https://github.com/dotnet/corefx/blob/c3053820eab4bf0fc9606dc6088c580ac3197201/src/System.Text.Json/tests/Resources/Payloads/BlogPost.json) generates the following POCO types:

https://gist.github.com/ahsonkhan/c92fa6f825849d7c8e1f8b3027754a50

The question is, where did the original JSON file itself come from?

Comment (ahsonkhan) [reply src/System.Text.Json/tests/Serialization/BlogPost.cs]: Also, for our particular use case here, I don't know if we need to use "real world" JSON blobs. We could generate our own sample .NET object types and use those to serialize large JSON payloads.

Comment (buyaa-n) [reply src/System.Text.Json/tests/Serialization/BlogPost.cs]: > I learned recently from @terrajobst, there is a way to generate .NET POCO from JSON built-into Visual Studio. I don't know how accurate it is, but we could consider using that instead.

I will consider using that

> The question is, where did the original JSON file itself come from

Not sure about that, anyways we are not using the json is that matter?

> Also, for our particular use case here of improving test coverage for large JSON, I don't know if we need to use "real world" JSON blobs. We could generate our own sample .NET object types and use those to serialize large JSON payloads.

Of course we can generate a large object for testing, that was easier way, but that would have some pattern/rule/loop which wouldn't be so close to or cover real life large objects scenario. So i preferred to use a POCO close to real

**Comment by buyaa-n** [comment]
@bartonjs @ahsonkhan comments addressed, please re-review

**Comment by buyaa-n** [comment]
@ahsonkhan @bartonjs please re review

File: src/System.Text.Json/tests/Serialization/Stream.WriteTests.cs [hunk 6]
```diff
@@ -258,6 +263,636 @@ public static async Task LargeJsonFile(int bufferSize)
...
+                                        {
+                                            new Alt
+                                            {
+                                                Type = 3,
+                                                Url = new Uri("http://test.com/link/entries/alts/1")
+                                            },
+                                            new Alt
+                                            {
+                                                Type = 2,
+                                                Url = new Uri("http://test.com/link/entries/alts/2")
```
Comment (bartonjs) [inline src/System.Text.Json/tests/Serialization/Stream.WriteTests.cs]: Artificial URLs should use a host from https://tools.ietf.org/html/rfc2606.

example.com, example.net, example.org, [anything].example, [anything].test, [anything].invalid.

File: src/System.Text.Json/tests/Serialization/Stream.WriteTests.cs [hunk 7]
```diff
@@ -258,6 +263,636 @@ public static async Task LargeJsonFile(int bufferSize)
...
+        private class FlushThresholdTestClass {
+            public string StringProperty { get; set; }
+            public List<int> ListOfInts { get; set; }
+            public FlushThresholdTestClass(List<int> list)
+            {
+                ListOfInts = list;
+            }
+        }
+
+        private static List<int> PopulateObjectWhichProduceJsonWithGivenSize(int size)
```
Comment (bartonjs) [inline src/System.Text.Json/tests/Serialization/Stream.WriteTests.cs]: "size" here seems ambiguous... I expected it to be the JSON output in bytes (or "runes"); but it's the number of elements in an array.  `GenerateListOfSize(int size)` seems better, perhaps?

File: src/System.Text.Json/tests/Serialization/Stream.WriteTests.cs [hunk 8]
```diff
@@ -258,6 +263,636 @@ public static async Task LargeJsonFile(int bufferSize)
...
+        [InlineData(65536)]
+        public static async void FlushThresholdTest(int bufferSize)
+        {
+            int thresholdSize = (int)(bufferSize * 0.9 - 2);
+            FlushThresholdTestClass serialaizeObject = new FlushThresholdTestClass(PopulateObjectWhichProduceJsonWithGivenSize(bufferSize));
+            List<object> list = new List<object>();
+            StringBuilder builder = new StringBuilder();
+            for (int i = 0; i < thresholdSize; i++)
+            {
+                builder.Append("a");
```
Comment (bartonjs) [inline src/System.Text.Json/tests/Serialization/Stream.WriteTests.cs]: This seems to just be building `new string('a', thresholdSize)`.

File: src/System.Text.Json/tests/Serialization/Stream.WriteTests.cs [hunk 9]
```diff
@@ -258,6 +263,636 @@ public static async Task LargeJsonFile(int bufferSize)
...
+        [InlineData(128)]
+        [InlineData(1024)]
+        [InlineData(4096)]
+        [InlineData(8192)]
+        [InlineData(16384)]
+        [InlineData(65536)]
+        public static async void FlushThresholdTest(int bufferSize)
+        {
+            int thresholdSize = (int)(bufferSize * 0.9 - 2);
+            FlushThresholdTestClass serialaizeObject = new FlushThresholdTestClass(PopulateObjectWhichProduceJsonWithGivenSize(bufferSize));
```
Comment (bartonjs) [inline src/System.Text.Json/tests/Serialization/Stream.WriteTests.cs]: NIt: Typo (the second 'a' doesn't belong).

File: src/System.Text.Json/tests/Serialization/Stream.WriteTests.cs [hunk 10]
```diff
@@ -258,6 +263,636 @@ public static async Task LargeJsonFile(int bufferSize)
...
+            {
+                builder.Append("a");
+            }
+            list.Add(builder.ToString());
+            serialaizeObject.StringProperty = builder.ToString();
+            list.Add(serialaizeObject);
+            JsonSerializerOptions options = new JsonSerializerOptions();
+            options.DefaultBufferSize = bufferSize;
+
+            string json = JsonSerializer.Serialize(list);
```
Comment (bartonjs) [inline src/System.Text.Json/tests/Serialization/Stream.WriteTests.cs]: I'm having trouble understanding what this test is testing.

* Given a size input, produce a list of that many `(int)1` values.
* Produce a string with approximately 90% as many ASCII letters in it as the List\<int\> has ints.
* Make an object which holds both the list and the string.
* Make a second list which has the string then the object.
* Serialize the second list

What aspect is being tested?  options.DefaultBufferSize is set, but if we're trying to measure that it has impact I don't see that being done here.

* Did we need to set DefaultBufferSize?
* If the list didn't have both the string and the object is the test substantially different?
* If the object didn't have both the string and the list of ints, is the test substantially different?

Based on the name, I'd expect something measuring that DefaultBufferSize is performing "as expected", but unless not setting it would throw (which doesn't seem to be the case, since the call with no options succeeds) I don't see how this test is testing it...  what would fail if the serializer never read that option at all?

If I'm completely missing what's being tested that's OK, but then we should come up with a combination of name changes and comments that make it apparent what's being tested and how it could fail.

File: src/System.Text.Json/tests/Serialization/Stream.WriteTests.cs (hunk 6 above)
Comment (buyaa-n) [reply src/System.Text.Json/tests/Serialization/Stream.WriteTests.cs]: will update

File: src/System.Text.Json/tests/Serialization/Stream.WriteTests.cs (hunk 10 above)
Comment (buyaa-n) [reply src/System.Text.Json/tests/Serialization/Stream.WriteTests.cs]: It is testing if serializer flushing buffer as expected for edge conditions like new object or array starts exactly at the buffer threshold (where buffer flush occurs), testing it with different buffer size => with different threshold size. It is covering following sentence from the issue description #39500 
image

File: src/System.Text.Json/tests/Serialization/Stream.WriteTests.cs [hunk 11]
```diff
@@ -258,6 +263,632 @@ public static async Task LargeJsonFile(int bufferSize)
...
+                    } 
+                };
+            
+                for (int j = 0; j < i % 4; j++)
+                {
+                    Paragraph purpleParagraph = new Paragraph
+                    {
+                        Name = $"a{i}b{j}",
+                        Type = i,
+                        Text = "The picture of two firefighters clearing up the snow surrounding a hydrant. In case there's an emergency, the cost in lives due to having snow blocking the hydrant is more significant than the cost of clearing them all beforehand. Clearing a hydrant after a snowfall is a problem worth solving.",
```
Comment (ahsonkhan) [inline src/System.Text.Json/tests/Serialization/Stream.WriteTests.cs]: We seem to be borrowing a lot of the "text" content from https://medium.com/@fagnerbrack/a-smart-programmer-understands-the-problems-worth-fixing-dcf15871f943

Can you just use dummy data instead? I think we should also be cautious when including anything that came specifically from the article (including links, image ids, etc.).

File: src/System.Text.Json/tests/Serialization/Stream.WriteTests.cs (hunk 5 above)
Comment (ahsonkhan) [reply src/System.Text.Json/tests/Serialization/Stream.WriteTests.cs]: Please add that as a comment within the test.

File: src/System.Text.Json/tests/Serialization/BlogPost.cs [hunk 12]
```diff
@@ -0,0 +1,364 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Net;
+using System.Text;
+
+namespace System.Text.Json.Serialization.Tests.Schemas.BlogPost
```
[the remark this answers is not in the document]
Comment (buyaa-n) [reply src/System.Text.Json/tests/Serialization/BlogPost.cs]: It was like that at first, updated because class names declared here very generic terms like `Post`, 'Member' can be overlapped easily https://github.com/dotnet/corefx/pull/39668#discussion_r306045839

File: src/System.Text.Json/tests/Serialization/Stream.WriteTests.cs [hunk 13]
```diff
@@ -258,6 +263,632 @@ public static async Task LargeJsonFile(int bufferSize)
...
+                posts[i - 1][0].Value.PrimaryTopic.RelatedTopics = posts[i];
+            }
+
+            JsonSerializerOptions options = new JsonSerializerOptions()
+            {
+                MaxDepth = depthFactor * 64,
+                IgnoreNullValues = ignoreNull,
+                WriteIndented = writeIndented
+            };
+            string json = JsonSerializer.Serialize(posts[0], options);
```
[the remark this answers is not in the document]
Comment (buyaa-n) [reply src/System.Text.Json/tests/Serialization/Stream.WriteTests.cs]: It is test for deep nested object. Array itself not used for serialization because first element within array is parent of all other array elements. i.e. 1st element is parent of 2nd, 2nd is parent of 3rd, 3rd is parent of 4th etc. Used array to keep reference to objects and assign next element as child of previous.  So even only first post from array used for serialization the array elements are used entirely in nested objects.

File: src/System.Text.Json/tests/Serialization/Stream.WriteTests.cs (hunk 11 above)
Comment (richlander) [reply src/System.Text.Json/tests/Serialization/Stream.WriteTests.cs]: I recommend [Project Gutenberg](http://www.gutenberg.org/) for large bodies of text.

Comment (buyaa-n) [reply src/System.Text.Json/tests/Serialization/Stream.WriteTests.cs]: Sure, links, names, images, ids updated/dummied, texts seemed to me dummy enough, but will update too

Comment (richlander) [reply src/System.Text.Json/tests/Serialization/Stream.WriteTests.cs]: The guidance is simple: don't take content from an unapproved resource. The whole of the web is that.

I've pre-approved two resources for the team and we can add others. For clarity, for content like this, anything that is public domain or has an appropriate creative commons license is fine. You'll have to show me the CC license if you want to use CC content.

File: src/System.Text.Json/tests/Serialization/Stream.WriteTests.cs (hunk 10 above)
Comment (buyaa-n) [reply src/System.Text.Json/tests/Serialization/Stream.WriteTests.cs]: Added comment explaining the object structure and its purpose in the test

``` 
// bufferSize * 0.9 is the threshold size from codebase, substruct 2 for [" characters, then create a 
// string containing (threshold - 2) amount of char 'a' which when written into output buffer produces buffer 
// which size equal to or very close to threshold size, then adding the string to the list, then adding a big 
// object to the list which changes depth of written json and should cause buffer flush
```
Hope its make it clear now and answering your first question.

>  Did we need to set DefaultBufferSize?

Yes to test with different threshold
> If the list didn't have both the string and the object is the test substantially different?

Yes as they have their own role, the string produces threshold length buffer when written, next object should cause buffer flush and its should be written without any loss  
> If the object didn't have both the string and the list of ints, is the test substantially different?

Not much, intended to test writing big enough object after flush

File: src/System.Text.Json/tests/System.Text.Json.Tests.csproj [hunk 14]
```diff
@@ -32,6 +32,7 @@
     <Compile Include="NewtonsoftTests\EnumConverterTests.cs" />
     <Compile Include="NewtonsoftTests\ImmutableCollectionsTests.cs" />
     <Compile Include="NewtonsoftTests\JsonSerializerTests.cs" />
+    <Compile Include="Serialization\BlogPost.cs" />
```
[the remark this answers is not in the document]
Comment (buyaa-n) [reply src/System.Text.Json/tests/System.Text.Json.Tests.csproj]: updated

File: src/System.Text.Json/tests/Serialization/BlogPost.cs [hunk 15]
```diff
@@ -0,0 +1,364 @@
...
+        public Value Value { get; set; }
+        public List<MentionedUser> MentionedUsers { get; set; }
+        public List<Collaborator> Collaborators { get; set; }
+        public bool HideMeter { get; set; }
+        public List<CollectionUserRelation> CollectionUserRelations { get; set; }
+        public object Mode { get; set; }
+        public References References { get; set; }
+    }
+
+    public class Value
```
[the remark this answers is not in the document]
Comment (buyaa-n) [reply src/System.Text.Json/tests/Serialization/BlogPost.cs]: added

File: src/System.Text.Json/tests/Serialization/Stream.WriteTests.cs (hunk 11 above)
Comment (buyaa-n) [reply src/System.Text.Json/tests/Serialization/Stream.WriteTests.cs]: How about adding this comment:
`// This schema populated using blogpost structure of https://medium.com`
https://github.com/dotnet/corefx/pull/39668/files#diff-d488e84455a3af1d990d5343131ade08R10
Any application talking to the site would need to produce json structure using their payload, so i assume its OK, used only schema and all data from payload dummied.

File: src/System.Text.Json/tests/Serialization/Stream.WriteTests.cs [hunk 16]
```diff
@@ -258,6 +263,636 @@ public static async Task LargeJsonFile(int bufferSize)
...
+
+            Assert.Throws<JsonException> (() => JsonSerializer.Serialize(posts[0], options));
+
+            using (var memoryStream = new MemoryStream())
+            {
+                await Assert.ThrowsAsync<JsonException>(async () => await JsonSerializer.SerializeAsync(memoryStream, posts[0], options));
+            }
+        }
+
+
```
Comment (ahsonkhan) [inline src/System.Text.Json/tests/Serialization/Stream.WriteTests.cs]: nit: spacing

File: src/System.Text.Json/tests/Serialization/SampleTestData.BlogPost.cs [hunk 17]
```diff
@@ -0,0 +1,367 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Collections.Generic;
+using System.Net;
+
+namespace System.Text.Json.Serialization.Tests.Schemas.BlogPost
+{ 
+// This schema populated using blogpost structure of https://medium.com
```
Comment (ahsonkhan) [inline src/System.Text.Json/tests/Serialization/SampleTestData.BlogPost.cs]: nit: spacing.

File: src/System.Text.Json/tests/Serialization/Stream.WriteTests.cs [hunk 18]
```diff
@@ -258,6 +263,636 @@ public static async Task LargeJsonFile(int bufferSize)
...
+                List<object> deserializedList = JsonSerializer.Deserialize<List<object>>(json, options);
+                Assert.Equal(stringOfThresholdSize, ((JsonElement)deserializedList[0]).GetString());
+                JsonElement.ObjectEnumerator obj = ((JsonElement)deserializedList[1]).EnumerateObject();
+                obj.MoveNext();               
+                Assert.Equal(stringOfThresholdSize, obj.Current.Value.GetString());
+            }
+        }
+
+        private class FlushThresholdTestClass {
```
Comment (ahsonkhan) [inline src/System.Text.Json/tests/Serialization/Stream.WriteTests.cs]: nit: formatting

File: src/System.Text.Json/tests/Serialization/Stream.WriteTests.cs [hunk 19]
```diff
@@ -258,6 +263,636 @@ public static async Task LargeJsonFile(int bufferSize)
...
             }
         }
+
+        [Theory]
+        [InlineData(1, true, true)]
+        [InlineData(1, true, false)]
+        [InlineData(1, false, true)]
+        [InlineData(1, false, false)]
+        [InlineData(10, true, false)]
+        [InlineData(10, false, false)]
```
Comment (ahsonkhan) [inline src/System.Text.Json/tests/Serialization/Stream.WriteTests.cs]: Is there a reason why we certain true/false combinations are chosen based on the payload size?

File: src/System.Text.Json/tests/Serialization/Stream.WriteTests.cs [hunk 20]
```diff
@@ -258,6 +263,636 @@ public static async Task LargeJsonFile(int bufferSize)
...
+        [Theory]
+        [InlineData(128)]
+        [InlineData(1024)]
+        [InlineData(4096)]
+        [InlineData(8192)]
+        [InlineData(16384)]
+        [InlineData(65536)]
+        public static async void FlushThresholdTest(int bufferSize)
+        {
+            // bufferSize * 0.9 is the threshold size from codebase, substruct 2 for [" characters, then create a
```
Comment (ahsonkhan) [inline src/System.Text.Json/tests/Serialization/Stream.WriteTests.cs]: nit: typo
```suggestion
            // bufferSize * 0.9 is the threshold size from codebase, subtract 2 for [" characters, then create a 
```

File: src/System.Text.Json/tests/Serialization/BlogPost.cs (hunk 12 above)
[the remark this answers is not in the document]
Comment (ahsonkhan) [reply src/System.Text.Json/tests/Serialization/BlogPost.cs]: Ah I see. That makes sense.

File: src/System.Text.Json/tests/Serialization/Stream.WriteTests.cs (hunk 19 above)
Comment (buyaa-n) [reply src/System.Text.Json/tests/Serialization/Stream.WriteTests.cs]: At first I was testing with all combinations, test became slower and producing very big json in writeIndented=true case, so decided to not use all combinations of options, especially for the bigger the payloads. Thought as writing without indentation works testing with indentation not much useful

Comment (ahsonkhan) [reply src/System.Text.Json/tests/Serialization/Stream.WriteTests.cs]: I see. How resource intensive are these new tests? And how long does it take to run them? If it is more than a couple of seconds, we should consider moving them to outerloop.

Also, do we need the ignore null option? It doesn't look like we have any nulls within the input object. If there is no difference between the null behaviors, it not necessary to test those combinations here.

Comment (buyaa-n) [reply src/System.Text.Json/tests/Serialization/Stream.WriteTests.cs]: > How resource intensive are these new tests? And how long does it take to run them? If it is more than a few of seconds, we should consider moving them to outerloop. I believe the entire set of JSON tests take ~15-20 seconds on my dev box, currently.

It became 20-22 seconds, so I decreased some tests test data including the above test. So now it is runs around 18 seconds on my machine, before adding these tests it was 16-17 seconds.

> Also, do we need the ignore null option? It doesn't look like we have any nulls within the input object. If there is no difference between the null behaviors for the objects being serialized/deserialized, its not necessary to test those combinations here.

There is null values, I just didn't explicitly set them into null, there were difference with running ignore null

File: src/System.Text.Json/tests/Serialization/Stream.WriteTests.cs (hunk 11 above)
Comment (richlander) [reply src/System.Text.Json/tests/Serialization/Stream.WriteTests.cs]: Sure. It is fine to use an API for a commercial service, like medium or Netflix. It is a problem if we store user content (like blog posts, in part or in whole) in tests because it is testing what the service returns. We may or may not  have to the rights to that, and I'm not interested in finding out in most cases (because it's a losing proposition, and there are better options).

I assume we're not actually going to call the medium service as part of our test runs. Is that right?

Comment (ahsonkhan) [reply src/System.Text.Json/tests/Serialization/Stream.WriteTests.cs]: > I assume we're not actually going to call the medium service as part of our test runs. Is that right?

Right.

**Comment by ericstj** [comment]
Looks like @buyaa-n addressed feedback, can folks please re-review?  I've kicked off another test run since previous one was a couple weeks old.

**Comment by buyaa-n** [comment]
@ahsonkhan @bartonjs please re review

File: src/System.Text.Json/tests/Serialization/SampleTestData.BlogPost.cs [hunk 21]
```diff
@@ -0,0 +1,367 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Collections.Generic;
+using System.Net;
+
+namespace System.Text.Json.Serialization.Tests.Schemas.BlogPost
+{ 
+    // This schema populated using blogpost structure of https://medium.com
```
Comment (stephentoub) [inline src/System.Text.Json/tests/Serialization/SampleTestData.BlogPost.cs]: Are there any copyright / proprietary / licensing concerns about doing so?

To be safe, I'd suggest instead basing it on something from bing or Azure or the dev blogs or some such thing, or even better just making something up.

Comment (buyaa-n) [reply src/System.Text.Json/tests/Serialization/SampleTestData.BlogPost.cs]: > Are there any copyright / proprietary / licensing concerns about doing so?

Not sure, not likely as any app would use the json structure to talk to this site
> To be safe, I'd suggest instead basing it on something from bing or Azure or the dev blogs or some such thing, or even better just making something up.

Sure I will look up for a big json schema from them and refactor the code

Comment (richlander) [reply src/System.Text.Json/tests/Serialization/SampleTestData.BlogPost.cs]: I already gave this feedback a couple months ago. Don't use content you find on the web in tests unless you know it is licensed for general use. Using blog content is even more challenging because there can be other posts on the same blog that are offensive (some of which might not be posted yet). It is very challenging to use content from medium and similar services. When in doubt, ask me or someone else who understands these issues first.

Feedback I gave earlier: https://github.com/dotnet/corefx/pull/39668#discussion_r314833357

Comment (ahsonkhan) [reply src/System.Text.Json/tests/Serialization/SampleTestData.BlogPost.cs]: To clarify, the tests don't use any "content" of any of the blogs published on medium. Only the schema/class definition is related to medium (i.e. the schema that would be required to load a JSON response from calling the medium REST API endpoint). Once we have the class, it's properties are manually populated with dummy data/content.

> Sure. It is fine to use an API for a commercial service, like medium or Netflix.

From your earlier comment, this PR is doing exactly that. The concerns that were brought up earlier were all related to the content itself (which have been addressed), and not the schema.

That said, I'd be OK with us not even using the medium developer APIs to generate classes (and use bing/azure/etc.), but I don't see it being absolutely necessary. I think there is value in testing popular/commonly used "real-world" end points and making sure the `JsonSerializer` can handle such classes/models. How can we avoid any references to those end-points in the tests if we want coverage for real scenarios. Is the suggestion to forgo that goal, and if not, what's the recommendation?

Comment (buyaa-n) [reply src/System.Text.Json/tests/Serialization/SampleTestData.BlogPost.cs]: Right, I have addressed @richlander's comment and updated the content, this is about the json schema/structure

> To be safe, I'd suggest instead basing it on something from bing or Azure or the dev blogs or some such thing, or even better just making something up.

By my understanding this is suggesting `just in case` use internal web/blog payloads or make up them. I am not planning to make up the schema as all other test schemas used in json tests are made up and with this one wanted to have one real world like scenario just like @ahsonkhan mentioned. So thinking to lookup payloads from internal sites and update as needed

Comment (richlander) [reply src/System.Text.Json/tests/Serialization/SampleTestData.BlogPost.cs]: We cannot use the schema of a commercial service unless there is a license that allows us to do so.

Comment (buyaa-n) [reply src/System.Text.Json/tests/Serialization/SampleTestData.BlogPost.cs]: I don't know, there is only info about contents nothing about public API payloads and don't think would find anything. Anyways as I said I will find payloads from `bing or Azure or the dev blogs` and use, as they are Microsoft product assume have no issue with licensing right?

Comment (richlander) [reply src/System.Text.Json/tests/Serialization/SampleTestData.BlogPost.cs]: Not exactly. This is an open source project, so using proprietary MS content is also a problem. Only use content that has a favorable license or terms of service and share a link with me to such a document.

Comment (buyaa-n) [reply src/System.Text.Json/tests/Serialization/SampleTestData.BlogPost.cs]: Sure will do, thanks

**Comment by maryamariyan** [comment]
Thank you for your contribution. As announced in dotnet/coreclr#27549 this repository will be moving to dotnet/runtime on November 13. If you would like to continue working on this PR after this date, the easiest way to move the change to dotnet/runtime is:
1.	In your corefx repository clone, create patch by running `git format-patch origin`
2.	In your runtime repository clone, apply the patch by running `git apply --directory src/corefx <path to the patch created in step 1>`

File: src/System.Text.Json/tests/Serialization/SampleTestData.BlogPost.cs (hunk 21 above)
Comment (buyaa-n) [reply src/System.Text.Json/tests/Serialization/SampleTestData.BlogPost.cs]: `bing or Azure or the dev blogs` doesn't seem to be open source. Even there is many knows open source blog/ecommerce platforms/CMSs, the web app built on them are not really open source, plus there is no any licensing/copyright info about how their Rest API payloads could be used. I was thinking to use Wikipedia as it is open source but didn't found info about using Rest API payload schema usage. I have talked about it with @richlander few days ago and API payload schema/structure licensing is such grey area we better not use in our code. So I am planning to just make up a schema of big object as @stephentoub suggested

File: src/System.Text.Json/tests/Serialization/SampleTestData.Order.cs [hunk 22]
```diff
@@ -0,0 +1,165 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Collections.Generic;
+
+namespace System.Text.Json.Serialization.Tests.Schemas.Order
+{ 
+    public class Order
```
Comment (stephentoub) [inline src/System.Text.Json/tests/Serialization/SampleTestData.Order.cs]: Nit: it's generally frowned upon having a class be in a namespace of the same name

File: src/System.Text.Json/tests/Serialization/SampleTestData.Order.cs [hunk 23]
```diff
@@ -0,0 +1,165 @@
...
+        public Address PickupAddress { get; set; }
+        public SampleEnumInt64 Coupon { get; set; }
+        public IEnumerable<Comment> UserInteractions { get; set; }
+        public DateTime Created { get; set; }
+        public DateTime Updated { get; set; }
+        public DateTime Confirmed { get; set; }
+        public DateTime ShippingDate { get; set; }
+        public DateTime EstimatedDelivery { get; set; }
+        public IEnumerable<Order> RelatedOrder { get; set; }
+		public User ReviewedBy { get; set; }
```
Comment (stephentoub) [inline src/System.Text.Json/tests/Serialization/SampleTestData.Order.cs]: Nit: spacing

File: src/System.Text.Json/tests/Serialization/SampleTestData.Order.cs [hunk 24]
```diff
@@ -0,0 +1,165 @@
...
+        public DateTime CreatedAt { get; set; }
+        public DateTime UpdatedAt { get; set; }
+        public string ImageId { get; set; }
+        public string TwitterId { get; set; }
+        public string FacebookId { get; set; }
+        public int SubscriptionType { get; set; }
+        public bool IsNew { get; set; }
+        public bool IsEmployee { get; set; }
+        public UserType UserType { get; set; }
+    }
```
Comment (stephentoub) [inline src/System.Text.Json/tests/Serialization/SampleTestData.Order.cs]: Nit: missing blank line after the brace

File: src/System.Text.Json/tests/Serialization/SampleTestData.Order.cs (hunk 22 above)
Comment (buyaa-n) [reply src/System.Text.Json/tests/Serialization/SampleTestData.Order.cs]: will update the namespace

File: src/System.Text.Json/tests/Serialization/Stream.WriteTests.cs [hunk 25]
```diff
@@ -2,8 +2,11 @@
 // The .NET Foundation licenses this file to you under the MIT license.
 // See the LICENSE file in the project root for more information.
 
+using System.Collections;
 using System.Collections.Generic;
 using System.IO;
+using System.Net;
```
Comment (ahsonkhan) [inline src/System.Text.Json/tests/Serialization/Stream.WriteTests.cs]: nit: Is this necessary? If not, remove unused using directive.

File: src/System.Text.Json/tests/Serialization/Stream.WriteTests.cs [hunk 26]
```diff
@@ -258,6 +263,368 @@ public static async Task LargeJsonFile(int bufferSize)
...
+                Assert.Equal(json, jsonSerialized);
+
+                memoryStream.Position = 0;
+                List<Order> deserializedList = await JsonSerializer.DeserializeAsync<List<Order>>(memoryStream, options);
+                Assert.Equal(payloadSize, deserializedList.Count);
+            }
+        }
+
+        [Theory]
+/*        [InlineData(1, true, true)] ActiveIssue https://github.com/dotnet/corefx/issues/41598
```
Comment (ahsonkhan) [inline src/System.Text.Json/tests/Serialization/Stream.WriteTests.cs]: @layomia - can you verify that these tests pass once your PR is in that fixes https://github.com/dotnet/corefx/issues/41598 (i.e. https://github.com/dotnet/corefx/issues/41598).

File: src/System.Text.Json/tests/Serialization/Stream.WriteTests.cs (hunk 25 above)
Comment (buyaa-n) [reply src/System.Text.Json/tests/Serialization/Stream.WriteTests.cs]: its not used now, will remove

File: src/System.Text.Json/tests/Serialization/Stream.WriteTests.cs [hunk 27]
```diff
@@ -258,6 +261,367 @@ public static async Task LargeJsonFile(int bufferSize)
                 Assert.Equal(SessionResponseCount, deserializedList.Count);
             }
         }
+
+        [Theory]
+//      [InlineData(1, true, true)]   ActiveIssue https://github.com/dotnet/corefx/issues/41598
```
Comment (ahsonkhan) [inline src/System.Text.Json/tests/Serialization/Stream.WriteTests.cs]: This issue has been fixed by https://github.com/dotnet/corefx/pull/42420. Can we re-enable these now?

**Review by ahsonkhan — APPROVED** [review]
Otherwise, looks good.

File: src/System.Text.Json/tests/Serialization/Stream.WriteTests.cs (hunk 27 above)
Comment (buyaa-n) [reply src/System.Text.Json/tests/Serialization/Stream.WriteTests.cs]: sure

**Comment by ahsonkhan** [comment]
Unrelated System.Net.Http.Functional.Tests failure: https://github.com/dotnet/corefx/issues/26539#issuecomment-553630592
